pkg/githubreconciler/reconciler.go

```diff
@@ -0,0 +1,126 @@
+/*
+Copyright 2025 Chainguard, Inc.
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package githubreconciler
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/go-github/v72/github"
+)
+
+// ReconcilerFunc is the function signature for GitHub resource reconcilers.
+// It receives the parsed resource information and appropriate GitHub client,
+// and returns an error if reconciliation fails.
+type ReconcilerFunc func(ctx context.Context, res *Resource, gh *github.Client) error
+
+// Resource represents a parsed GitHub resource (issue or pull request).
```

Comment:
There are lots of event types we might care about, e.g., pushes, and CheckRun events (retries, etc.)

We already support lots of those events in github-events, and have BQ recording set up with schemas. Can we reuse any of that here?

Reply:
See my comment above, but it's not such a simple thing to generically map events to workqueues.

Think about all of the helper functions that exist to do this mapping in K8s:
1. Watching our own resource type
2. Watching child resource types and chasing OwnerRefs
3. Leveraging things like the Knative `tracker` to watch ObjectReferences.
...

Reply:
I buy that it's not easy to do in general, I just wanted to reuse as much as we have already rather than only building from scratch.

In a workqueue-centric model, do we not still want to record all the events, and get support for types we already have a use case for?

If we didn't compose the two, do you envision that we should separately subscribe to GH events to get them into BQ, and handle CheckRun retry webhooks via PubSub and issue updates via WQ? I hope not.